orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/Stage.java

```diff
@@ -0,0 +1,232 @@
+package com.netflix.spinnaker.orca.pipeline.model;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.batch.StageBuilder;
+import com.netflix.spinnaker.orca.listeners.StageTaskPropagationListener;
+import com.netflix.spinnaker.orca.pipeline.util.StageNavigator;
+import org.codehaus.groovy.runtime.ReverseListIterator;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+
+public interface Stage<T extends Execution<T>> {
+  String getRefId();
+
+  void setRefId(String refId);
+
+  /**
+   * A stage's unique identifier
+   */
+  String getId();
+
+  /**
+   * The type as it corresponds to the Mayo configuration
+   */
+  String getType();
+
+  /**
+   * The name of the stage. Can be different from type, but often will be the same.
+   */
+  String getName();
+
+  /**
+   * Gets the execution object for this stage
+   */
+  T getExecution();
+
+  /**
+   * Gets the start time for this stage. May return null if the stage has not been started.
+   */
+  Long getStartTime();
+
+  /**
+   * Gets the end time for this stage. May return null if the stage has not yet finished.
+   */
+  Long getEndTime();
+
+  void setStartTime(Long startTime);
+
+  void setEndTime(Long endTime);
+
+  /**
+   * The execution status for this stage
+   */
+  ExecutionStatus getStatus();
+
+  /**
+   * sets the execution status for this stage
+   */
+  void setStatus(ExecutionStatus status);
+
+  /**
+   * Gets the last stage preceding this stage that has the specified type.
+   */
+  default Stage preceding(String type) {
+    int i = getExecution()
```

Comment:
This is still a poor method.

If anything delegating to `ancestors()` seems more appropriate (provided ancestors handles the situation where there are no refIds on stages).